src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/SelectExpandNode.cs

```diff
@@ -242,30 +253,125 @@ public SelectExpandNode(SelectExpandClause selectExpandClause, IEdmStructuredTyp
+                            if (!ExpandedProperties.ContainsKey(navigationProperty))
+                            {
+                                ExpandedProperties.Add(navigationProperty, expandItem);
+                            }
+                            else
+                            {
+                                ExpandedProperties[navigationProperty] = expandItem;
+                            }
                         }
                         else
                         {
                             ReferencedNavigationProperties.Add(navigationProperty);
                         }
                     }
+                    else
+                    {
+                        //This is the case where the navigation property is not on the current type. We need to propagate the expand item to deeper SelectExpandNode.
+                        IEdmProperty complexProperty = FindNextPropertySegment(expandReferenceItem.PathToNavigationProperty);
+
+                        if (complexProperty != null && ExpandedNavigationPropertiesOnComplexTypes.ContainsKey(complexProperty))
+                        {
+                            SelectExpandClause oldClause = ExpandedNavigationPropertiesOnComplexTypes[complexProperty].SelectAndExpand;
+                            SelectExpandClause newClause = new SelectExpandClause(oldClause.SelectedItems.Concat(new List<SelectItem>() { expandReferenceItem }), false);
+                            ExpandedNavigationSelectItem newItem = new ExpandedNavigationSelectItem(expandReferenceItem.PathToNavigationProperty, navigationSegment.NavigationSource, newClause);
+                            ExpandedNavigationPropertiesOnComplexTypes.Add(complexProperty, newItem);
+                        }
+                        else if (complexProperty != null)
+                        {
+                            List<SelectItem> expandedItemsForClause = new List<SelectItem>();
+                            expandedItemsForClause.Add(expandReferenceItem);
+                            SelectExpandClause clause = new SelectExpandClause(expandedItemsForClause, false);
+                            ExpandedNavigationSelectItem expandItem = new ExpandedNavigationSelectItem(expandReferenceItem.PathToNavigationProperty, navigationSegment.NavigationSource, clause);
+                            ExpandedNavigationPropertiesOnComplexTypes.Add(complexProperty, expandItem);
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Finds the appropriate property segment which should be responsible for propagating the expand item.
+        /// For instance, if we are creating the SelectExpandNode for property p2 for the query ~/EnitySet/Key/p1/p2/p3?$expand=NP, we want to return property p3 here. 
+        /// </summary>
+        private IEdmStructuralProperty FindNextPropertySegment(ODataPath path)
```

Comment:
>FindNextPropertySegment [](start = 39, length = 23)

Would like to have you walk me through this and explain your assumptions. i.e., it looks like it will ignore properties in the path that inconveniently don't match what's in the PathToProperty?

Reply:
Understand the logic after walking through it with you, and it's fine given the current design. We will probably need to modify a bit once we do select options on complex types, at which time we can revisit how properties of complex types in $select are handled in general and ideally make it more similar to how we handle nested entities.

---
In reply to: [308497764](https://github.com/OData/WebApi/pull/1738#discussion_r308497764) [](ancestors = 308497764)